// ng-katex/src/ng-katex.directive.js
import { Directive, ElementRef, EventEmitter, Input, Output, } from '@angular/core';
import { KatexService } from './ng-katex.service';
var KatexDirective = /** @class */ (function () {
    function KatexDirective(el, katexService) {
        this.el = el;
        this.katexService = katexService;
        this.onError = new EventEmitter();
    }
    KatexDirective.prototype.ngOnChanges = function () {
        try {
            this.katexService.render(this.equation, this.el, this.options);
        }
        catch (e) {
            this.onError.emit(e);
        }
    };
    KatexDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[katex]',
                },] },
    ];
    /** @nocollapse */
    KatexDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: KatexService, },
    ]; };
    KatexDirective.propDecorators = {
        'equation': [{ type: Input, args: ['katex',] },],
        'options': [{ type: Input, args: ['katex-options',] },],
        'onError': [{ type: Output },],
    };
    return KatexDirective;
}());
export { KatexDirective };
